### DISCUSSION TITLE
Incorrect calculations

### DISCUSSION THREAD

**Discussion by tcjw3** [WillCodeForCats/solaredge-modbus-multi#550] [General]
Hello WillCodeForCats,

A great job, thank you very much, a huge help to integrate the PV system from Solaredge into HomeAssistant!
However, I believe that an error has crept in:
You are using some sensors that originally have different units:
"name: "Power - Grid Import"                                             
        unit_of_measurement: "W" "
is not correct, as the sensor "solaredge_m1_ac_power" used outputs the unit kW. As a result, subsequent calculations with this sensor are incorrect if they are carried out with a sensor with the original unit "W".
It is necessary to multiply all sensors with the original unit "kW" by 1000 so that the calculations are correct.
This applies to:
sensor.solaredge_m1_ac_power

This can also cause subsequent calculations to be incorrect:
z. B.
"name: "Power - PV - Load"                                               
        unit_of_measurement: "W"                                                
        device_class: "power"
        state_class: "measurement"                                              
        # What is being generated less what is going to the battery and what is being exported, always positive.
        state: "{{ max([float(states('sensor.power_solar_generation'), 0) - float(states('sensor.power_pv_battery'), 0) - float(states('sensor.power_grid_export'), 0), 0]) }}"

The previously created sensor "sensor.power_grid_export" has assumed the unit W, but is too low by a factor of 1000 because it had the unit kW in the original sensor.

This means that all sensors that originate from calculations with the "sensor.solaredge_m1_ac_power" sensor are currently not calculated correctly.

I think this needs to be changed in the yaml:
{{ min([float(states('sensor.solaredge_m1_ac_power'), 0), 0]) | abs() * 1000 }}

Am I perhaps right with my comment?

Best regards
Thomas

**Comment by WillCodeForCats** [comment]
I haven't really looked at the templates since I don't use them personally. But I would say make sure the sensors are set to the correct unit the template is expecting, or update the template to the units you are using.

Sensor entities can be customized between different units in the supported device class:

https://www.home-assistant.io/integrations/sensor/

https://www.home-assistant.io/docs/configuration/customizing-devices/

**Reply by tcjw3** [reply]
You are absolutely right. I'm still a bit of a newbie here. I hadn't taken that into account. But with the photo instruction from WebSpider, I've now understood it too. Thanks again for your work and the current tips.
Thomas

**Comment by tcjw3** [comment]
``  - sensor:                                                                     
      - name: "Power - Battery Charging"
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"                                              
        state: "{{ max([float(states('sensor.solaredge_b1_dc_power'), 0), 0]) | abs() }}"

      - name: "Power - Battery Discharging"                                     
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"                                              
        state: "{{ min([float(states('sensor.solaredge_b1_dc_power'), 0), 0]) | abs() }}"

      - name: "Power - Grid Import"                                             
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"
        icon: mdi:transmission-tower-export
        state: "{{ min([float(states('sensor.solaredge_m1_ac_power'), 0), 0]) | abs() * 1000 }}"

      - name: "Power - Grid Export"                                             
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"
        icon: mdi:transmission-tower-import
        state: "{{ max([float(states('sensor.solaredge_m1_ac_power'), 0), 0]) | abs() * 1000 }}"

      - name: "Power - Solar Generation"                                        
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"
        icon: mdi:solar-power
        # The net of DC power from inverter and battery, only if this is positive
        state: "{{ max([float(states('sensor.solaredge_i1_dc_power'), 0) + float(states('sensor.solaredge_b1_dc_power'), 0), 0]) }}"

      - name: "Power - Consumption"                                             
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"                                              
        # The net of inverter and grid power                                    
        # state: "{{ float(states('sensor.solaredge_i1_ac_power'), 0) - float(states('sensor.solaredge_m1_ac_power'), 0) 0] }}"
        state: "{{ float(states('sensor.solaredge_i1_ac_power') | default(0)) - float(states('sensor.solaredge_m1_ac_power') | default(0)) }}"

      - name: "Power - PV - Battery"                                            
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"                                              
        # Battery power less anything coming from the grid, always positive     
        state: "{{ max([float(states('sensor.solaredge_b1_dc_power'), 0) + min([float(states('sensor.solaredge_i1_dc_power'), 0), 0]), 0]) }}"

      - name: "Power - PV - Load"                                               
        unit_of_measurement: "W"                                                
        device_class: "power"
        state_class: "measurement"                                              
        # What is being generated less what is going to the battery and what is being exported, always positive.
        state: "{{ max([float(states('sensor.power_solar_generation'), 0) - float(states('sensor.power_pv_battery'), 0) - float(states('sensor.power_grid_export'), 0), 0]) }}"

      - name: "Power - PV - Grid"                                               
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"                                              
        # What is being generated less what is going to the battery and what is being consumed, always positive
        state: "{{ max([float(states('sensor.power_solar_generation'), 0) - float(states('sensor.power_pv_battery'), 0) - float(states('sensor.power_consumption'), 0), 0]) }}"

      - name: "Power - Grid - Load"                                             
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"                                              
        # What is being imported less anything going to the battery, always positive
        state: "{{ max([float(states('sensor.power_grid_import'), 0) + min([float(states('sensor.solaredge_i1_ac_power'), 0), 0]), 0]) }}"

      - name: "Power - Grid - Battery"                                          
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"                                              
        # Grid -> Battery is I1 AC power, when negative                         
        # Battery -> Grid is what is being exported, if the battery is discharging
        state: "{{ max([-float(states('sensor.solaredge_i1_ac_power'), 0), 0]) - (float(states('sensor.power_grid_export'), 0) if float(states('sensor.solaredge_b1_dc_power'), 0) < 0 else 0) }}"

      - name: "Power - Battery - Load"                                          
        unit_of_measurement: "W"                                                
        device_class: "power"                                                   
        state_class: "measurement"                                              
        # What is being discharged minus what is being exported, if positive    
        state: "{{ max([float(states('sensor.power_battery_discharging'), 0) - float(states('sensor.power_grid_export'), 0), 0]) }}"
        
  - sensor:
      - name: "Solar Panel Production W"
        unit_of_measurement: "W"
        icon: mdi:solar-power
        state: >
          {% set i1_dc_power = states('sensor.solaredge_i1_dc_power') | float(0) %}
          {% set b1_dc_power = states('sensor.solaredge_b1_dc_power') | float(0) %}
          {% if (i1_dc_power + b1_dc_power <= 0) %}
            0
          {% else %}
            {{ (i1_dc_power + b1_dc_power) }}
          {% endif %}
        availability: >
          {{ states('sensor.solaredge_i1_dc_power') | is_number and states('sensor.solaredge_b1_dc_power') | is_number }}

      - name: "Solar House Consumption W"
        unit_of_measurement: "W"
        icon: mdi:home
        state: >
          {% set i1_ac_power = states('sensor.solaredge_i1_ac_power') | float(0) %}
          {% set m1_ac_power_kW = states('sensor.solaredge_m1_ac_power') | float(0) %}
          {% set m1_ac_power_W = m1_ac_power_kW * 1000 %}  {# Convert kW to W #}
          {% if (i1_ac_power - m1_ac_power_W <= 0) %}
            0
          {% else %}
            {{ (i1_ac_power - m1_ac_power_W) }}
          {% endif %}
        availability: >
          {{ states('sensor.solaredge_i1_ac_power') | is_number and states('sensor.solaredge_m1_ac_power') | is_number }}

      - name: "Solar Inverter Consumption w"
        unit_of_measurement: "W"
        icon: mdi:home-lightning-bolt-outline
        state: >
          {% set i1_ac_power = states('sensor.solaredge_i1_ac_power') | float(0) %}
          {% set i1_dc_power = states('sensor.solaredge_i1_dc_power') | float(0) %}
          {% if (i1_ac_power > i1_dc_power < 0) %}
            {{ (i1_ac_power - i1_dc_power)| round(2) }}
          {% else %}
            {{ ( i1_dc_power - i1_ac_power)| round(2)}}
          {% endif %}
        availability: >
          {{ states('sensor.solaredge_i1_ac_power') | is_number and states('sensor.solaredge_i1_dc_power') | is_number }}

      - name: "Solar Imported Power W"
        unit_of_measurement: "W"
        icon: mdi:transmission-tower-export
        state: >
          {% set m1_ac_power_kW = states('sensor.solaredge_m1_ac_power') | float(0) %}
          {% set m1_ac_power_W = m1_ac_power_kW * 1000 %}          
          {% if (m1_ac_power_W < 0) %}
            {{ (m1_ac_power_W * -1) }}
          {% else %}
            0
          {% endif %}
        availability: >
          {{ states('sensor.solaredge_m1_ac_power') | is_number}}

      - name: "Solar Exported Power W"
        unit_of_measurement: "W"
        icon: mdi:transmission-tower-import
        state: >
          {% set m1_ac_power_kW = states('sensor.solaredge_m1_ac_power') | float(0) %}
          {% set m1_ac_power_W = m1_ac_power_kW * 1000 %}  
          {% if (m1_ac_power_W > 0) %}
            {{ (m1_ac_power_W) }}
          {% else %}
            0
          {% endif %}
        availability: >
          {{ states('sensor.solaredge_m1_ac_power') | is_number }}

      - name: "Energy - Battery Content"
        unit_of_measurement: "kWh"
        device_class: "energy"
        state_class: "total"
        state: "{{ (float(states(('sensor.solaredge_b1_state_of_energy'))) * 13.8 / 100) | round (2) }}"        

          
sensor:
  - platform: integration
    source: sensor.solar_panel_production_w
    method: left
    unit_prefix: k
    name: solar_panel_production_kwh
  - platform: integration
    source: sensor.solar_house_consumption_w
    method: left
    unit_prefix: k
    name: solar_house_consumption_wh
  - platform: integration
    source: sensor.solar_imported_power_w
    method: left
    unit_prefix: k
    name: solar_imported_power_kwh
  - platform: integration
    source: sensor.solar_exported_power_w
    method: left
    unit_prefix: k
    name: solar_exported_power_kwh
    
utility_meter:
  solar_panel_production_hourly:
    source: sensor.solar_panel_production_kwh
    name: Solar Panel Production hourly
    cycle: hourly
  solar_house_consumption_hourly:
    source: sensor.solar_house_consumption_kwh
    name: Solar House Consumption hourly
    cycle: hourly
  solar_imported_power_hourly:
    source: sensor.solar_imported_power_kwh
    name: Solar Imported Power hourly
    cycle: hourly
  solar_exported_power_hourly:
    source: sensor.solar_exported_power_kwh
    name: Solar Exported Power hourly
    cycle: hourly

  solar_panel_production_daily:
    source: sensor.solar_panel_production_kwh
    name: Solar Panel Production Daily
    cycle: daily
  solar_house_consumption_daily:
    source: sensor.solar_house_consumption_kwh
    name: Solar House Consumption Daily
    cycle: daily
  solar_imported_power_daily:
    source: sensor.solar_imported_power_kwh
    name: Solar Imported Power Daily
    cycle: daily
  solar_exported_power_daily:
    source: sensor.solar_exported_power_kwh
    name: Solar Exported Power Daily
    cycle: daily

  solar_imported_power_weekly:
    source: sensor.solar_imported_power_kwh
    name: Solar Imported Power Weekly
    cycle: weekly
  solar_house_consumption_weekly:
    source: sensor.solar_house_consumption_kwh
    name: Solar House Consumption Weekly
    cycle: weekly
  solar_panel_production_weekly:
    source: sensor.solar_panel_production_kwh
    name: Solar Panel Production Weekly
    cycle: weekly
  solar_exported_power_weekly:
    source: sensor.solar_exported_power_kwh
    name: Solar Exported Power Weekly
    cycle: weekly
    
    
  solar_imported_power_monthly:
    source: sensor.solar_imported_power_kwh
    name: Solar Imported Power Monthly
    cycle: monthly
  solar_house_consumption_monthly:
    source: sensor.solar_house_consumption_kwh
    name: Solar House Consumption Monthly
    cycle: monthly
  solar_panel_production_monthly:
    source: sensor.solar_panel_production_kwh
    name: Solar Panel Production Monthly
    cycle: monthly
  solar_exported_power_monthly:
    source: sensor.solar_exported_power_kwh
    name: Solar Exported Power Monthly
    cycle: monthly
    
  solar_imported_power_yearly:
    source: sensor.solar_imported_power_kwh
    name: Solar Imported Power Yearly
    cycle: yearly
  solar_house_consumption_yearly:
    source: sensor.solar_house_consumption_kwh
    name: Solar House Consumption Yearly
    cycle: yearly
  solar_panel_production_yearly:
    source: sensor.solar_panel_production_kwh
    name: Solar Panel Production Yearly
    cycle: yearly
  solar_exported_power_yearly:
    source: sensor.solar_exported_power_kwh
    name: Solar Exported Power Yearly
    cycle: yearly

**Comment by tcjw3** [comment]
I have multiplied the sensor twice by 1000 and performed an intermediate operation several times:
z. B. 
 - name: "Solar Exported Power W"
        unit_of_measurement: "W"
        icon: mdi:transmission-tower-import
        state: >
          {% set m1_ac_power_kW = states('sensor.solaredge_m1_ac_power') | float(0) %}
          {% set m1_ac_power_W = m1_ac_power_kW * 1000 %}  
          {% if (m1_ac_power_W > 0) %}
            {{ (m1_ac_power_W) }}
          {% else %}
            0
          {% endif %}
        availability: >
          {{ states('sensor.solaredge_m1_ac_power') | is_number }}

All measured values and egg units are now

Best regards, Tom

**Reply by WebSpider** [reply]
It's a lot easier to let HA do the unit conversion:

![image](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/437323/8b434d7c-cf5b-45c4-847d-dbd7b2ff4865)

**Reply by tcjw3** [reply]
ok, that might seem easier

**Reply by tcjw3** [reply]
But you really could have told me that earlier. How stupid can you be? (I mean me)

**Comment by mpredfearn** [comment]
The templates are on a Wiki page - just edit it?

**Reply by tcjw3** [reply]
I have edited it in my HA energy.yaml and it works.
